lib/runBlock.js

```diff
@@ -32,237 +32,187 @@ module.exports = function (opts, cb) {
     return cb(new Error('invalid input, block must be provided'))
   }
 
-  const self = this
+  _runBlock.bind(this)(opts)
+    .then((results) => cb(null, results))
+    .catch(cb)
+}
 
-  // parse options
+async function _runBlock (opts) {
+  const state = new PStateManager(this.stateManager)
   const block = opts.block
-  const skipBlockValidation = opts.skipBlockValidation || false
   const generateStateRoot = !!opts.generate
-  const validateStateRoot = !generateStateRoot
-  const bloom = new Bloom()
-  const receiptTrie = new Trie()
-  // the total amount of gas used processing this block
-  var gasUsed = new BN(0)
-  var checkpointedState = false
-  var receipts = []
-  var txResults = []
-  var result
 
-  // run everything
-  async.series([
-    beforeBlock,
-    validateBlock,
-    setStateRoot,
-    checkpointState,
-    processTransactions,
-    payOmmersAndMiner
-  ], parseBlockResults)
+  /**
+   * The `beforeBlock` event
+   *
+   * @event Event: beforeBlock
+   * @type {Object}
+   * @property {Block} block emits the block that is about to be processed
+   */
+  await this._emit('beforeBlock', opts.block)
 
-  function beforeBlock (cb) {
-    /**
-     * The `beforeBlock` event
-     *
-     * @event Event: beforeBlock
-     * @type {Object}
-     * @property {Block} block emits the block that is about to be processed
-     */
-    self.emit('beforeBlock', opts.block, cb)
+  // Set state root if provided
+  if (opts.root) {
+    await state.setStateRoot(opts.root)
   }
 
-  function afterBlock (cb) {
-    /**
-     * The `afterBlock` event
-     *
-     * @event Event: afterBlock
-     * @type {Object}
-     * @property {Object} result emits the results of processing a block
-     */
-    self.emit('afterBlock', result, cb)
+  // Checkpoint state
+  await state.checkpoint()
```

Comment:
``setStateRoot``, ``checkpointState`` logic ok, just for reference: order of ``validateBlock`` has been changed respectively moved further down the line.